packages/language/src/language-server/document-symbol-builder.ts

```diff
@@ -0,0 +1,326 @@
+ *
+ * Copyright Contributors to the Zowe Project.
+ *
+ */
+
+import {
+  DocumentSymbol,
+  SymbolKind,
+  Position,
+} from "vscode-languageserver-types";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { DeclaredItem, Statement, SyntaxKind } from "../syntax-tree/ast";
+import { Range, rangeToLSP, getSyntaxNodeRange, tokenToRange } from "./types";
+import { IToken } from "chevrotain";
+import { CstNodeKind } from "../syntax-tree/cst";
+interface LevelSymbol {
+  level: number | null;
+  symbol: DocumentSymbol;
+}
+
+interface SymbolBuilder {
+  canHandle(token: IToken): boolean;
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[];
+}
+
+class ProcedureSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    return (
+      token.tokenTypeIdx == CstNodeKind.EntryStatement_Variable &&
+      token.payload?.element.kind === SyntaxKind.ProcedureStatement
+    );
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const labelPrefixStatement = token.payload?.element.container as Statement;
+    const procedureName =
+      labelPrefixStatement.labels.map((label) => label.name).join(" ") ??
+      "<unknown>";
+    const range = getSyntaxNodeRange(labelPrefixStatement.labels[0]) ?? {
+      start: 0,
+      end: 0,
+    };
+
+    const symbol = createDocumentSymbol(
+      procedureName,
+      SymbolKind.Function,
+      range,
+      childSymbols,
+      textDocument,
+    );
+
+    symbol.range.end = rangeToLSP(
+      textDocument,
+      tokenToRange(elementTokens[elementTokens.length - 1]),
+    ).end;
+    return [symbol];
+  }
+}
+
+class DeclareSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    if (
+      token.tokenTypeIdx !== CstNodeKind.DefineOrdinalStatement_Unsigned0 ||
+      token.payload?.element.kind !== SyntaxKind.DeclareStatement
+    ) {
+      return false;
+    }
+
+    // There is still a chance that all variables are generated by the preprocessor.
+    // However, to check this, we would need to traverse the remaining subtree, so we
+    // do the check during symbol creation. No invalid symbols will be created.
+    return true;
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const documentSymbols: DocumentSymbol[] = [];
+    const levelSymbols: LevelSymbol[] = [];
+    const hierarchyBuilder = new LevelHierarchyBuilder();
+
+    for (const item of token.payload?.element.items.filter(
+      (i: DeclaredItem) => i.kind === SyntaxKind.DeclaredItem,
+    )) {
+      levelSymbols.push(
+        ...this.retrieveLevelSymbols(item, null, childSymbols, textDocument),
+      );
+    }
```

Comment:
Suggestion: We should try to keep the amount of `any` to a minimum. Try to cast the `token.payload.element` object first, so we get better IDE support.